test/ir-to-ir/Tests/LiftProgramLambdasSpec.hs

```diff
@@ -0,0 +1,84 @@
+module Tests.LiftProgramLambdasSpec where
+
+import qualified IR.IR                         as I
+import qualified IR.Types.Classes              as C
+
+import IR.LambdaLift
+import           Common.Pretty                  ( spaghetti )
+import           Common.Compiler                ( runPass )
+import           Front.Ast
+import           Front.Parser                   ( parseProgram )
+import           Front.ParseOperators           ( parseOperators )
+import           IR.LowerAst                    ( lowerProgram )
+import IR.ClassInstantiation (instProgram)
+import IR.Types.Poly as Poly
+import           IR.TypeInference               ( inferExpr
+                                                , inferProgram
+                                                )
+
+import           Control.Comonad                ( Comonad(..) )
+import           Data.Bifunctor                 ( Bifunctor(second) )
+import           Test.Hspec                     ( Spec(..)
+                                                , it
+                                                , pending
+                                                , shouldBe
+                                                , describe
+                                                )
+
+lowerAndLift :: Either String Program -> Either String (I.Program Poly.Type)
+lowerAndLift (Left e) = Left "Failed to parse program"
+lowerAndLift (Right p) =
+  case runPass $ lowerProgram (parseOperators p) >>= inferProgram >>= instProgram >>= liftProgramLambdas of
+    Left e' -> Left $ show e'
+    Right p' -> Right p'
+
+spec :: Spec
+spec = do
+
+  it "lifts a lambda without free variables" $ do
+    let nonLiftedProgram = parseProgram $ unlines
+          [ "bar: Int = 5"
+          , "baz x: Int -> Int ="
+          , "  x + 1"
+          , "foo y: Int -> Int ="
+          , "  let adder (z: Int) -> Int = z + 1"
+          , "  adder y"
+          ]
+        liftedProgram = parseProgram $ unlines
+          [ "bar: Int = 5"
+          , "baz x: Int -> Int ="
+          , "  x + 1"
+          , "anon0 z: Int -> Int ="
+          , "  z + 1"
+          , "foo y: Int -> Int ="
+          , "  let adder = anon0"
+          , "  adder y"
+          ]
+        nestedToLifted = lowerAndLift nonLiftedProgram
+        liftedToLifted = lowerAndLift liftedProgram
+    nestedToLifted `shouldBe` liftedToLifted
+
+  it "lifts a lambda with free variables" $ do
+    let nonLiftedProgram = parseProgram $ unlines
+          [ "bar: Int = 5"
+          , "baz x: Int -> Int ="
+          , "  x + 1"
+          , "foo y: Int -> Int ="
+          , "  let w = 1"
+          , "      adder (z: Int) -> Int = z + bar + w"
+          , "  adder y"
+          ]
+        liftedProgram = parseProgram $ unlines
+          [ "bar: Int = 5"
+          , "baz x: Int -> Int ="
+          , "  x + 1"
+          , "anon0 (w: Int) (z: Int) -> Int ="
```

Comment:
fyi @j-hui as I was writing this, I discovered that the compiler doesn't like `anon0 w z: Int -> Int -> Int`. Any idea why? I get this error:

`TypeError "Cannot change empty Ann type to Classes type"`

Should I open a separate issue?

Reply:
Interesting. This error is generated inside type inference and is caused by no/incomplete type annotation of lambda functions. I tried to print out the lowered AST of this example to see how the type annotations look like at that stage (i.e., right before type inference), and I got:
```
    ( anon0
    , Lambda
      (Just x)
      (Lambda
        (Just z)
        (Prim
          (PrimOp PrimAdd)
          [ Prim (PrimOp PrimAdd) [Var x (Type []), Var z (Type [])] (Type [])
          , Lit (LitIntegral 1) (Type [])
          ]
          (Type [])
        )
        (Type [TBuiltin (Arrow (Type []) (Type []))])
      )
      (Type
        [ TBuiltin
          (Arrow
            (Type [TBuiltin (Integral 32)])
            (Type
              [ TBuiltin
                  (Arrow (Type [TBuiltin (Integral 32)])
                         (Type [TBuiltin (Integral 32)])
                  )
              ]
            )
          )
        , TBuiltin
          (Arrow (Type []) (Type [TBuiltin (Arrow (Type []) (Type []))]))
        ]
      )
    )
```
It seems like even though we have annotated the types of both inputs, somehow it get lost. Currently the type inference on main branch still requires that the input of lambda functions has its type annotated, so thus we got this error.

Reply:
That's weird because `baz x: Int -> Int` works fine.

![image](https://user-images.githubusercontent.com/22177160/143803953-df8a40ef-9aae-4e93-9e22-a272ca34fcf2.png)

Reply:
Ah I think the problem is inside `lowerProgram`, we failed to pass along the type information when we recursively `lowerLambda`. Can you please try to change the `id` in [this line](https://github.com/ssm-lang/sslang/blob/main/src/IR/LowerAst.hs#L165) to `k` and see if it works? I honestly don't really know what's going on inside `LowerAST` LOL.

Reply:
I don't think that line is the issue (and I don't see what we could change it to either).

The information isn't being lost either; it's present in the first type annotation of the entire lambda term, `Int -> Int -> Int`. That should be unified with the second, less specific annotation `_ -> _ -> _`. 

We need to obtain a trace to figure out exactly what IR term is causing this error.

Reply:
Sorry, you are right -- I misunderstood the `k` here. Changing `id` to `k` will simply append `Int -> Int -> Int` to the type annotation of inner lambda (`z -> z + bar + w`), which is incorrect. My bad.

Reply:
Was this ever resolved?

> We need to obtain a trace to figure out exactly what IR term is causing this error.

Did we ever figure out what was causing the issue?

Reply:
Opened #62